transaction.go

```diff
@@ -49,198 +90,303 @@ type Transaction struct {
-	t.firstPage.pageStatus = pageStatusComitted
-
-	// calculate the checksum and write it to the last page
-	checksum, err := t.checksum()
-	if err != nil {
-		return build.ExtendErr("Unable to create checksum of transaction", err)
-	}
-	t.finalPage.transactionChecksum = checksum
-	err = t.finalPage.Write(t.wal.logFile)
-	if err != nil {
-		return build.ExtendErr("Writing the final page failed", err)
+// NewTransaction creates a transaction from a set of updates
+func (w *WAL) NewTransaction(updates []Update) *Transaction {
+	// Create new transaction
+	newTransaction := Transaction{
+		Updates:      updates,
+		wal:          w,
+		initComplete: make(chan error),
 	}
 
-	// Finalize the commit by writing the first page with the updated status
-	err = t.firstPage.Write(t.wal.logFile)
-	if err != nil {
-		return build.ExtendErr("Writing the first page failed", err)
-	}
+	// Initialize the transaction by splitting up the payload among free pages
+	// and writing them to disk.
+	go threadedInitTransaction(&newTransaction)
 
-	t.commitComplete = true
-	t.wal.logFile.Sync()
-	return nil
+	return &newTransaction
 }
 
-// SignalApplyComplete  informs the WAL that it is safe to free the used pages to reuse them in a new transaction
-func (t *Transaction) SignalApplyComplete() <-chan error {
+// SignalUpdatesApplied  informs the WAL that it is safe to free the used pages to reuse them in a new transaction
+func (t *Transaction) SignalUpdatesApplied() <-chan error {
 	if !t.setupComplete || !t.commitComplete || t.releaseComplete {
-		panic("misuse of transaction - call each of the signaling methods exactly ones, in serial, in order")
+		panic("misuse of transaction - call each of the signaling methods exactly once, in serial, in order")
 	}
 	t.releaseComplete = true
 	notifyChannel := make(chan error)
 
 	go func() {
-		t.wal.mu.Lock()
-		defer t.wal.mu.Unlock()
-
 		// Set the page status to applied
 		t.firstPage.pageStatus = pageStatusApplied
 
 		// Write the page to disk
-		err := t.firstPage.Write(t.wal.logFile)
+		var err error
+		if t.wal.deps.disrupt("ReleaseFail") {
+			// Disk failure causes the commit to fail
+			err = errors.New("Write failed on purpose")
+		} else {
+			err = t.firstPage.Write(t.wal.logFile)
+		}
+
 		if err != nil {
 			notifyChannel <- build.ExtendErr("Couldn't write the page to file", err)
 			return
 		}
-
+		t.wal.fSync()
 		// Update the wallets available pages
 		page := t.firstPage
+		t.wal.mu.Lock()
 		for page != nil {
-			// Calculate and append the index of the page
+			// Append the index of the freed page
 			t.wal.availablePages = append(t.wal.availablePages, page.offset)
 			page = page.nextPage
 		}
+		t.wal.mu.Unlock()
 		notifyChannel <- nil
 	}()
 	return notifyChannel
 }
+
+// SignalSetupComplete will signal to the WAL that any required setup has
+// completed, and that the WAL can safely commit to the transaction being
+// applied atomically.
+func (t *Transaction) SignalSetupComplete() <-chan error {
```

Comment:
I think a better technique here would be to make this function fully blocking instead of launching another thread. If the caller doesn't want to block, the caller can call `go SignalSetupComplete()`. The docstring for this function would then need to clarify that it is a blocking call which performs an fsync, and therefore is pretty slow.

Reply:
ah whoops, meant to place this on the `SignalUpdatesApplied` function, ignore the above comment for this function.